alg/sha256.c

```diff
@@ -391,6 +396,14 @@ SHA256_Init(SHA256_CTX * ctx)
 
 	/* Initialize state. */
 	memcpy(ctx->state, initial_state, sizeof(initial_state));
+
+	/* Choose hardware acceleration (if any). */
+	if (hwaccel == HW_UNSET) {
```

Comment:
Is this faster than placing

```
if (hwaccel != HW_UNSET)
    return;
```

into the top of `hwaccel_init`?  I think it's cleaner if we the `hwaccel_init` and `assert` are unconditional here, but not so much cleaner to be worth paying a performance penalty.

Reply:
Doesn't make any difference.  We only call SHA256_Init once in the libcperciva test suite and the spiped proto.  It's called a few times in scrypt's `scryptenc_setup` for the header checksum and signature, and in `scryptdec_setup` to verify the same -- but again, that's a function that's only run once.

tarsnap's `crypto_hash_data_key_2` and `crypto_hash_data_2` does call _Init every time, but I haven't seen those (as a time blocker) in any profiling.